src/unitxt/metrics.py

```diff
@@ -400,31 +699,180 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
+            subgroup_field: name of field in instance additional_inputs that contains the subgroup identifier
+            baseline_name: value of subgroup_field that indicates a score belongs in baseline_scores.
+
+        Returns:
+            List of dicts, each corresponding to a group of instances (defined by grouping_field)
+        """
+        from collections import defaultdict
+
+        # two-level defaultdict: first is the grouping, second is the field name
+        # first list is instance scores for baseline group, second is for comparison group
+        group_to_instance_scores = defaultdict(lambda: defaultdict(lambda: [[], []]))
 
         for instance in instances:
-            yield instance
+            additional_inputs = instance["additional_inputs"]
+            for cc in [self.grouping_field, subgroup_field]:
+                if cc not in additional_inputs:
+                    raise ValueError(
+                        f"Missing '{cc}' from instance {instance}. "
+                        f"This field is required for group based metric computation."
+                    )
+            group_key = additional_inputs[self.grouping_field]
+            # indicator if is in the baseline group or not
+            is_baseline = str(additional_inputs[subgroup_field]) == baseline_name
+            # convert True (baseline) to 0, and False (others) to 1, store in respective groups
+            idx = int(not is_baseline)
+
+            for field_name in field_names:
+                group_to_instance_scores[group_key][field_name][idx].append(
+                    instance["score"]["instance"][field_name]
+                )
+
+        # now for each group, take the aggregation function, comparing others to baseline
+        return [
+            {
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(
+                            baseline_scores=scores[field_name][0],
+                            other_scores=scores[field_name][1],
+                        )
+                        for field_name in field_names
+                    }
+                }
+            }
+            for scores in group_to_instance_scores.values()
+        ]
+
+    def _set_up_group_mean_aggregation(
+        self, instances, reduction_params, reduction_fields
+    ):
+        # if treat groups as units
+        if reduction_params["agg_func"][2]:
+            # pass the group aggregate---not instance---scores to resample as usual
+            aggregation_function = self.average_instance_scores
+            instances_to_resample = self.get_group_aggregated_instance_scores(
+                instances, reduction_fields, reduction_params["agg_func"][1]
+            )
+        else:
+            # pass the instance scores to resample, and calculate the group aggregation on the resamplings
+            instances_to_resample = deepcopy(instances)
```

Comment:
why is the copy needed?

Reply:
Because in the global calculation I want to simply pass one object, instances_to_resample (or scores_to_resample), which differs based on whether the grouping is applied.  In once case it is simply the instances, in other cases it is the group score applied.